ModuleInterface: Avoid printing `@_spi` enum elements in public swiftinterfaces

tshortli opened the pull request:
The `@_spi` attribute was accepted on enum case declarations but the cases were still printed in public swiftinterfaces, leaking the SPI. The compiler now avoids printing enum cases in public swiftinterfaces when they are declared as `@_spi` by checking whether the enum case declaration contains any enum elements that are SPI (enum case declarations apparently do not themselves have a copy of the attributes they were written with). The compiler now also diagnoses `@_spi` on enum case declarations when the enum is `@frozen` since all the cases of a frozen enum must be known to the client so that the memory layout for the enum can be computed.

This PR also includes tests that show that the exportability of case statements is not checked. Unfortunately this gap is not simple to close since exportability is usually checked during availability checking but availability checking isn't appropriate for case statements and there isn't existing infrastructure to check exportability independently. A follow up PR will be needed to address this.

Resolves rdar://72873771

tshortli committed:
ModuleInterface: Avoid printing `@_spi` enum case decls in public swiftinterfaces.

rdar://72873771

tshortli committed:
Sema: Diagnose `@_spi` on elements of `@frozen` enums.

rdar://72873771

tshortli committed:
NFC: Increase test coverage for SPI enum elements.

These tests identified a gap in type checking the exportability of case statements. Unfortunately this gap is not simple to close, since exportability is usually checked during availability checking but availability checking isn't appropriate for case statements and there isn't existing infrastructure to check exportability independently.

tshortli commented:
@swift-ci please test

tshortli commented on test/SPI/resilience.swift:
```diff
@@ -1,29 +0,0 @@
-// Test limits of SPI members in frozen types.
```
This file was deleted because the tests it contains seem to be completely superseded by the `spi_members.swift` test.

tshortli commented on test/SPI/private_swiftinterface.swift:
```diff
@@ -34,14 +34,6 @@
 // RUN: %target-swift-frontend -typecheck-module-from-interface -I %t %t/Main.swiftinterface
 // RUN: %target-swift-frontend -typecheck-module-from-interface -I %t %t/Main.private.swiftinterface -module-name Main
 
-/// Serialize and deserialize this module, then print.
```
I removed this because it interfered with the stability of printing the interface with an enum case declaration like `case a, b`. It seemed safe to remove to me because there are many other tests exercising serialization and deserialization of the `@_spi` attribute simply by compiling a helper module with `@_spi` declarations. If this is still needed for some reason, though, I can split it into a new test case.

nkcsgexi commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -276,6 +276,21 @@ PrintOptions PrintOptions::printSwiftInterfaceFile(ModuleDecl *ModuleToPrint,
+          // Enum elements are usually not printed, so we have to override the
+          // print option controlling that. We only check the first element
+          // because all the elements in a single case decl should have the same
+          // characteristics.
```
Should we assert that all the elements return the same `shouldPrint` result? otherwise, LGTM!

tshortli replied:
I think that if were to do this kind of validation I would rather create some helper APIs on `EnumCaseDecl` to codify that using any element is appropriate for this kind of check and then assert during parsing that we parse all elements with the same attributes. This approach of grabbing the first case is already used in a couple of disparate locations so asserting this here seems insufficiently general.

tshortli replied:
I've added a `getFirstElement()` utility and adopted it throughout to codify this practice.

nkcsgexi replied:
Thank you!

tshortli committed:
AST: Introduce a `getFirstElement()` utility on `EnumCaseDecl`.

This utility helps codify the practice of using the attributes of the first element of a case decl as the attributes for the entire case.

tshortli commented:
@swift-ci please test
